impl/src/main/java/com/okta/sdk/impl/oauth2/AccessTokenRetrieverServiceImpl.java

```diff
@@ -0,0 +1,249 @@
+/*
+ * Copyright 2020-Present Okta, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.okta.sdk.impl.oauth2;
+
+import com.okta.commons.http.MediaType;
+import com.okta.commons.http.authc.DisabledAuthenticator;
+import com.okta.commons.lang.Assert;
+import com.okta.sdk.client.AuthenticationScheme;
+import com.okta.sdk.client.AuthorizationMode;
+import com.okta.sdk.impl.api.OAuth2TokenClientCredentialsResolver;
+import com.okta.sdk.impl.config.ClientConfiguration;
+import com.okta.sdk.impl.error.DefaultError;
+import com.okta.sdk.resource.ExtensibleResource;
+import com.okta.sdk.resource.ResourceException;
+import io.jsonwebtoken.Jwts;
+import org.bouncycastle.asn1.pkcs.PrivateKeyInfo;
+import org.bouncycastle.openssl.PEMKeyPair;
+import org.bouncycastle.openssl.PEMParser;
+import org.bouncycastle.openssl.jcajce.JcaPEMKeyConverter;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.io.Reader;
+import java.nio.charset.Charset;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.security.InvalidKeyException;
+import java.security.KeyPair;
+import java.security.PrivateKey;
+import java.time.Instant;
+import java.time.temporal.ChronoUnit;
+import java.util.Date;
+import java.util.UUID;
+
+/**
+ * Implementation of {@link AccessTokenRetrieverService} interface.
+ *
+ * This has logic to fetch OAuth2 access token from the Authorization server endpoint.
+ *
+ * @since 1.6.0
+ */
+public class AccessTokenRetrieverServiceImpl implements AccessTokenRetrieverService {
+    private static final Logger log = LoggerFactory.getLogger(AccessTokenRetrieverServiceImpl.class);
+
+    private static final String TOKEN_URI  = "/oauth2/v1/token";
+
+    private ClientConfiguration tokenClientConfiguration;
+    private OAuth2TokenClient tokenClient;
```

Comment:
nit: these should be final
(**NOTE:** make sure you have your IDE configure to detect these things)